File: app/src/main/java/app/kamy/qalbuApp/infrastructure/preferences/DhikrStore.kt
package app.kamy.qalbuApp.infrastructure.preferences

import android.content.Context

data class DhikrPreset(
    val id: String,
    val labelResKey: String,
    val arabic: String,
    val target: Int
)

object DhikrStore {
    private const val PREFS = "qalbu_dhikr"
    private const val KEY_COUNT_PREFIX = "count_"
    private const val KEY_TOTAL_PREFIX = "total_"

    val presets: List<DhikrPreset> = listOf(
        DhikrPreset("subhanallah", "dhikr_subhanallah", "سُبْحَانَ اللَّهِ", 33),
        DhikrPreset("alhamdulillah", "dhikr_alhamdulillah", "الْحَمْدُ لِلَّهِ", 33),
        DhikrPreset("allahuakbar", "dhikr_allahuakbar", "اللَّهُ أَكْبَرُ", 34),
        DhikrPreset("istighfar", "dhikr_istighfar", "أَسْتَغْفِرُ اللَّهَ", 100),
        DhikrPreset("salawat", "dhikr_salawat", "اللَّهُمَّ صَلِّ عَلَى مُحَمَّد", 100)
    )

    fun sessionCount(context: Context, presetId: String): Int =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getInt(KEY_COUNT_PREFIX + presetId, 0)

    fun totalCount(context: Context, presetId: String): Int =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getInt(KEY_TOTAL_PREFIX + presetId, 0)

    fun increment(context: Context, presetId: String): Int {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val next = prefs.getInt(KEY_COUNT_PREFIX + presetId, 0) + 1
        prefs.edit()
            .putInt(KEY_COUNT_PREFIX + presetId, next)
            .putInt(KEY_TOTAL_PREFIX + presetId, prefs.getInt(KEY_TOTAL_PREFIX + presetId, 0) + 1)
            .apply()
        return next
    }

    fun resetSession(context: Context, presetId: String) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putInt(KEY_COUNT_PREFIX + presetId, 0)
            .apply()
    }
}

File: iosApp/Saat/Infrastructure/Persistence/Local/DhikrStore.swift
//
//  DhikrStore.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

internal struct DhikrPreset: Identifiable, Sendable {
    internal let id: String
    internal let arabic: String
    internal let target: Int
    
    internal let labelEn: String
    internal let labelId: String
    
    internal let translitEn: String
    internal let translitId: String
    
    internal let meaningEn: String
    internal let meaningId: String
    
    internal func label(for language: String) -> String {
        return (language == "id" || language == "ms") ? labelId : labelEn
    }
    
    internal func translit(for language: String) -> String {
        return (language == "id" || language == "ms") ? translitId : translitEn
    }
    
    internal func meaning(for language: String) -> String {
        return (language == "id" || language == "ms") ? meaningId : meaningEn
    }
}

internal final class DhikrStore: Sendable {
    private static let prefsSuite = "saat_dhikr"
    private static let keyCountPrefix = "count_"
    private static let keyTotalPrefix = "total_"
    
    private static var defaults: UserDefaults {
        return UserDefaults(suiteName: prefsSuite) ?? UserDefaults.standard
    }
    
    internal static let presets: [DhikrPreset] = [
        DhikrPreset(
            id: "subhanallah",
            arabic: "سُبْحَانَ اللَّهِ",
            target: 33,
            labelEn: "SubhanAllah",
            labelId: "SubhanAllah",
            translitEn: "Subḥānallāh",
            translitId: "Subḥānallāh",
            meaningEn: "Glory be to Allah",
            meaningId: "Maha Suci Allah"
        ),
        DhikrPreset(
            id: "alhamdulillah",
            arabic: "الْحَمْدُ لِلَّهِ",
            target: 33,
            labelEn: "Alhamdulillah",
            labelId: "Alhamdulillah",
            translitEn: "Al-ḥamdu lillāh",
            translitId: "Al-ḥamdu lillāh",
            meaningEn: "All praise is for Allah",
            meaningId: "Segala puji bagi Allah"
        ),
        DhikrPreset(
            id: "allahuakbar",
            arabic: "اللَّهُ أَكْبَرُ",
            target: 34,
            labelEn: "Allahu Akbar",
            labelId: "Allahu Akbar",
            translitEn: "Allāhu akbar",
            translitId: "Allāhu akbar",
            meaningEn: "Allah is the Greatest",
            meaningId: "Allah Maha Besar"
        ),
        DhikrPreset(
            id: "istighfar",
            arabic: "أَسْتَغْفِرُ اللَّهَ",
            target: 100,
            labelEn: "Istighfar",
            labelId: "Istighfar",
            translitEn: "Astaghfirullāh",
            translitId: "Astaghfirullāh",
            meaningEn: "I seek forgiveness from Allah",
            meaningId: "Aku mohon ampun kepada Allah"
        ),
        DhikrPreset(
            id: "salawat",
            arabic: "اللَّهُمَّ صَلِّ عَلَى مُحَمَّد",
            target: 100,
            labelEn: "Salawat",
            labelId: "Salawat",
            translitEn: "Allāhumma ṣalli ʿalā Muḥammad",
            translitId: "Allāhumma ṣalli ʿalā Muḥammad",
            meaningEn: "O Allah, send blessings upon Muhammad",
            meaningId: "Ya Allah, limpahkan shalawat kepada Muhammad"
        )
    ]
    
    internal static func sessionCount(for presetId: String) -> Int {
        return defaults.integer(forKey: keyCountPrefix + presetId)
    }
    
    internal static func totalCount(for presetId: String) -> Int {
        return defaults.integer(forKey: keyTotalPrefix + presetId)
    }
    
    internal static func increment(for presetId: String) -> Int {
        let currentSession = sessionCount(for: presetId)
        let nextSession = currentSession + 1
        let currentTotal = totalCount(for: presetId)
        
        defaults.set(nextSession, forKey: keyCountPrefix + presetId)
        defaults.set(currentTotal + 1, forKey: keyTotalPrefix + presetId)
        
        return nextSession
    }
    
    internal static func resetSession(for presetId: String) {
        defaults.set(0, forKey: keyCountPrefix + presetId)
    }
}
